internal/proxyservice/proxy_service.go

```diff
@@ -57,78 +59,246 @@ func (s *proxyServices) Stop() error {
+//   gets written, so I'm making it public.
+type HTTPServiceConfig struct {
+	SharedListenOn    string
+	SubserviceConfigs []v2.Service
+}
 
-	tcpPlugins := s.availPlugins.TCPPlugins()
-	// httpPlugins := s.availPlugins.HTTPPlugins()
-
-	for _, svc := range s.config.Services {
-		requestedPlugin := svc.Connector //TODO: this rename is a name smell
-
-		// first check the available TCP Plugins
-		tcpPlugin, found := tcpPlugins[requestedPlugin]
-		if found {
-			tcpSvc, err := s.createTCPService(svc, tcpPlugin)
-			// Failure to create is fatal.
-			if err != nil {
-				// TODO: Add Fatalf to our logger and use that
-				s.logger.Panicf("unable to create TCP service '%s'", svc.Name)
-			}
-			servicesToStart = append(servicesToStart, tcpSvc)
+//func NewHTTPService(listenOn string) *HTTPServiceConfig {
+//	return &HTTPServiceConfig{
+//		SharedListenOn: listenOn,
+//		SubserviceConfigs: []*v2.Service{},
+//	}
+//}
+//
+//func (hs *HTTPServiceConfig) AddSubService(svc *v2.Service) {
+//	hs.SubserviceConfigs = append(hs.SubserviceConfigs, svc)
+//}
+
+// configsByType holds proxy service configuration in a form that directly
+// corresponds to the ProxyService objects we want to create.  One ProxyService
+// will be created for each entry in http, and one for each entry in tcp.
+type configsByType struct {
+	tcp  []v2.Service
+	http []HTTPServiceConfig
+}
+
+// newConfigsByType converts a slice of v2.Service configs into the configs
+// needed to actually created ProxyServices -- configsByType.  In particular, it
+// takes all the http configs and creates proper HTTPServiceConfig objects out
+// of them -- grouping the raw v2.Service configs by their listenOn property.
+// The remaining services are tcp, and already correspond to the services we'll
+// run.
+// TODO: Eventually it should get moved to the config package, and this
+//   transformation should occur at the border.
+func newConfigsByType(
+	uncheckedConfigs []*v2.Service,
+	availPlugins plugin.AvailablePlugins,
+) configsByType {
+
+	// Get the nil checks out of the way.
+	// TODO: This Config definition itself should be changed to []v2.Service
+	var rawConfigs []v2.Service
+	for _, cfg := range uncheckedConfigs {
+		if cfg == nil {
+			// Hard-coding log here is okay since nils should never occur and
+			// we won't be unit testing this.  Instead, we'll make the change
+			// in the TODO_ above and this code will be deleted then.
+			log.Fatalln("Nil configuration is not allowed!")
+		}
+		rawConfigs = append(rawConfigs, *cfg)
+	}
+
+	// Separate HTTP and TCP configs
+	var httpConfigs []v2.Service
+	var	tcpConfigs []v2.Service
+
+	// TODO: Add proper validation here of the type.  This should moved into
+	//   IsHTTPPlugin, whose API will likely change to returning a type or an
+	//   error
+	for _, cfg := range rawConfigs {
+		if plugin.IsHTTPPlugin(availPlugins, cfg.Connector) {
+			httpConfigs = append(httpConfigs, cfg)
```

Comment:
It seems that we are leaking implementation details up to the level where they might not need to be visible - I think the knowledge of how to squash and parse http service configs probably needs to be in some `.... .http ...` package since the `proxyservice` should ideally not need to know anything other than how to start/stop services

Reply:
This is exactly right.  You'll note the comment at the top of the method :)

> // TODO: Eventually it should get moved to the config package, and this
> //   transformation should occur at the border.